include/nbl/builtin/hlsl/ieee754.hlsl

```diff
@@ -0,0 +1,137 @@
+	NBL_CONSTEXPR_STATIC_INLINE bit_rep_t min = 1ull << base_t::mantissaBitCnt;
+	NBL_CONSTEXPR_STATIC_INLINE int exponentMax = exponentBias;
+	NBL_CONSTEXPR_STATIC_INLINE int exponentMin = -(exponentBias - 1);
+};
+
+template <typename T>
+inline uint32_t extractBiasedExponent(T x)
+{
+	using AsUint = typename unsigned_integer_of_size<sizeof(T)>::type;
+	return glsl::bitfieldExtract<AsUint>(ieee754::impl::bitCastToUintType(x), traits<typename float_of_size<sizeof(T)>::type>::mantissaBitCnt, traits<typename float_of_size<sizeof(T)>::type>::exponentBitCnt);
+}
+
+template<>
+inline uint32_t extractBiasedExponent(uint64_t x)
+{
+	uint64_t output = (x >> traits<float64_t>::mantissaBitCnt) & (traits<float64_t>::exponentMask >> traits<float64_t>::mantissaBitCnt);
+	return uint32_t(output);
+}
+
+template<>
+inline uint32_t extractBiasedExponent(float64_t x)
+{
+	return extractBiasedExponent<uint64_t>(ieee754::impl::bitCastToUintType(x));
+}
+
+template <typename T>
+inline int extractExponent(T x)
+{
+	using AsFloat = typename float_of_size<sizeof(T)>::type;
+	return int(extractBiasedExponent(x)) - traits<AsFloat>::exponentBias;
+}
+
+template <typename T>
+NBL_CONSTEXPR_INLINE_FUNC T replaceBiasedExponent(T x, typename unsigned_integer_of_size<sizeof(T)>::type biasedExp)
+{
+	using AsFloat = typename float_of_size<sizeof(T)>::type;
+	return impl::castBackToFloatType<T>(glsl::bitfieldInsert(ieee754::impl::bitCastToUintType(x), biasedExp, traits<AsFloat>::mantissaBitCnt, traits<AsFloat>::exponentBitCnt));
+}
+
+// performs no overflow tests, returns x*exp2(n)
+template <typename T>
+NBL_CONSTEXPR_INLINE_FUNC T fastMulExp2(T x, int n)
+{
+	return replaceBiasedExponent(x, extractBiasedExponent(x) + uint32_t(n));
+}
+
+template <typename T>
+NBL_CONSTEXPR_INLINE_FUNC typename unsigned_integer_of_size<sizeof(T)>::type extractMantissa(T x)
+{
+	using AsUint = typename unsigned_integer_of_size<sizeof(T)>::type;
+	return ieee754::impl::bitCastToUintType(x) & traits<typename float_of_size<sizeof(T)>::type>::mantissaMask;
+}
+
+template <typename T>
+NBL_CONSTEXPR_INLINE_FUNC typename unsigned_integer_of_size<sizeof(T)>::type extractNormalizeMantissa(T x)
+{
+	using AsUint = typename unsigned_integer_of_size<sizeof(T)>::type;
+	using AsFloat = typename float_of_size<sizeof(T)>::type;
+	return extractMantissa(x) | (AsUint(1) << traits<AsFloat>::mantissaBitCnt);
+}
+
+template <typename T>
+NBL_CONSTEXPR_INLINE_FUNC typename unsigned_integer_of_size<sizeof(T)>::type extractSign(T x)
+{
+	using AsFloat = typename float_of_size<sizeof(T)>::type;
+	return (ieee754::impl::bitCastToUintType(x) & traits<AsFloat>::signMask) >> ((sizeof(T) * 8) - 1);
+}
+
+template <typename T>
+NBL_CONSTEXPR_INLINE_FUNC typename unsigned_integer_of_size<sizeof(T)>::type extractSignPreserveBitPattern(T x)
+{
+	using AsFloat = typename float_of_size<sizeof(T)>::type;
+	return ieee754::impl::bitCastToUintType(x) & traits<AsFloat>::signMask;
+}
+
```

Comment:
@keptsecret here's where we'll add `T flipSign(T, bool)` and `T setSign(T,bool)`

Reply:
resolving